tokio/src/io/join.rs

```diff
@@ -0,0 +1,122 @@
+//! Join two values implementing `AsyncRead` and `AsyncWrite` into a single one.
+
+use crate::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+cfg_io_util! {
+    pin_project_lite::pin_project! {
+        /// Joins two values implementing `AsyncRead` and `AsyncWrite` into a
+        /// single handle.
+        pub struct Join<R, W> {
+            #[pin]
+            reader: R,
+            #[pin]
+            writer: W,
+        }
+    }
+
+    impl<R, W> Join<R, W>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite + Unpin,
+    {
+        /// Join two values implementing `AsyncRead` and `AsyncWrite` into a
+        /// single handle.
+        pub fn new(reader: R, writer: W) -> Self {
+            Self { reader, writer }
+        }
+
+        /// Splits this `Join` back into its `AsyncRead` and `AsyncWrite`
+        /// components.
+        pub fn split(self) -> (R, W) {
+            (self.reader, self.writer)
+        }
+
+        /// Returns a reference to the inner reader.
+        pub fn reader(&self) -> &R {
+            &self.reader
+        }
+
+        /// Returns a reference to the inner writer.
+        pub fn writer(&self) -> &W {
+            &self.writer
+        }
+
+        /// Returns a mutable reference to the inner reader.
+        pub fn reader_mut(&mut self) -> &mut R {
+            &mut self.reader
+        }
+
+        /// Returns a mutable reference to the inner writer.
+        pub fn writer_mut(&mut self) -> &mut W {
+            &mut self.writer
+        }
```

Comment:
I would probably also add `reader_pin_mut`/`writer_pin_mut` with this signature:
```
pub fn reader_pin_mut(self: Pin<&mut Self>) -> Pin<&mut R>;
```